sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,750 @@
+`dim0 ...` are the values of the dimensions from the
+`[[reqd_work_group_size()]]` attribute, in the same order as they appear in the
+attribute.
+
+**NOTE**: One may wonder why the type of the "reqd\_sub\_group\_size" property
+is not `PI_PROPERTY_TYPE_UINT32` since its value is always 32-bits.  The
+reason is that we may want to expand this property in the future to contain a
+list of required sub-group sizes.  Likewise, the "reqd\_work\_group\_size"
+property may be expanded in the future to contain a list of required work-group
+sizes.
+
+
+### Changes specific to AOT mode
+
+In AOT mode, for each AOT target specified by the `-fsycl-targets` command
+line option, DPC++ normally invokes the AOT compiler for each device IR module
+resulting from the sycl-post-link tool. For example, this is `ocloc` command
+for Intel Gen AOT target and `opencl-aot` command for the x86 AOT target with
+SPIR-V as the input, or other specific tools for the PTX target with LLVMIR
+bitcode input. This causes a problem, though, for IR modules that use optional
+features because these commands could fail if they attempt to compile IR using
+an optional feature that is not supported by the target device.  We therefore
+need some way to avoid calling these commands in these cases.
+
+The overall design is as follows.  The DPC++ installation includes a
+configuration file that has one entry for each device that it supports. Each
+entry lists the set of aspects that the device supports and a list of the
+sub-group sizes that it supports.  DPC++ then consults this configuration
+file to decide whether to invoke a particular AOT compiler on each device IR
+module, using the information from the module's "SYCL/image-requirements"
+property set.
+
+#### Device configuration file
+
+The configuration file uses a simple YAML format where each top-level key is
+a name of a device architecture. These names correspond to SYCL aspect enum
+identifiers as defined in the [TBD] API header. There are sub-keys under each
+device for the supported aspects, sub-group sizes and AOT compiler ID.  For
+example:
+
+```
+gen11_1:
+  aspects: [1, 2, 3]
+  sub-group-sizes: [8, 16]
+  aot-compiler-id: gen-spir64
+gen_icl:
+  aspects: [2, 3]
+  sub-group-sizes: [8, 16]
+  aot-compiler-id: gen-spir64
+x86_64_avx512:
+  aspects: [1, 2, 3, 9, 11]
+  sub-group-sizes: [8, 32]
+  aot-compiler-id: x86-spir64
+```
+
+The values of the aspects in this configuration file can be the numerical
+values from the `enum class aspect` enumeration or the enum identifer itself.
+For each valid AOT compiler ID the driver has a built-in rule how to construct
+an AOT compilation command line based on given architecture name. For example,
+for the `gen11_1` and `gen_icl` architectures, the driver sees `gen-spir64`
+as the AOT compiler ID, so it knows that the `ocloc` tool must be used, and it
+also knows how to translate the `gen11_1` or `gen_icl` to proper `ocloc`
+architecture specification option.
```

Comment:
The advantage is that we decouple architecture (stem) from the AOT compiler. With it we may map any architecture to any AOT compiler. For Intel AOT compilers we likely don't need this, I agree. I don't have strong preference for `aot-compiler-id`. 

Another note is that the actual user-friendly architectures in `-fsycl-targets` should look like gen_<sub_arch>, I don't think we need to repeat spir64 in it. It should map 1:1 to aspect enum identifiers.

Reply:
> The advantage is that we decouple architecture (stem) from the AOT compiler. With it we may map any architecture to any AOT compiler. For Intel AOT compilers we likely don't need this, I agree. I don't have strong preference for aot-compiler-id.

I think I see your intent, but it seems like it has limited value because the compiler driver still needs to have built-in knowledge about how to construct the command line for the AOT command.  I think this could be useful if the config file abstracted away all the knowledge about invoking the AOT command.  For example:

```
gen11_1:
  aspects: [1, 2, 3]
  sub-group-sizes: [8, 16]
  aot-compiler-command: ocloc -file ${SPIRV} -spirv_input -output ${BINARY} -output_no_suffix -device gen11
```

This would abstract away all knowledge of the command line from the compiler driver.  It just needs to substitute the `${SPIRV}` and `${BINARY}` pseudo-variables.  However, I realize that's a lot more work.  If we wanted to do this, it seems like it might be better as a separate design document.

In the short term, it seems better to me to eliminate `aot-compiler-id` until we can make it more general.

Reply:
let's start with no `aot-compiler-id`, can be added later if needed w/o breaking backward compatibility

Reply:
> aot-compiler-command: ocloc -file ${SPIRV} -spirv_input -output ${BINARY} -output_no_suffix -device gen11

This seems like a possibly significant security problem.  I am pretty sure I wouldn't want the driver execing off some arbitrary command that the user can control via this file.

Reply:
Removed in 722e4ca76464489ee4062209e481d212d33aeef2

Reply:
> This seems like a possibly significant security problem. I am pretty sure I wouldn't want the driver execing off some arbitrary command that the user can control via this file.

I agree! However, is this roughly the same concern as a makefile's ability to spawn arbitrary commands as part of the user's build system, or is it worse for some reason?

Reply:
IMO, it is worse, because the driver and this architecture is designed by Intel, and thus we will be responsible for any security issues arising from it.